src/Libraries/Microsoft.Extensions.Http.Resilience/Hedging/Internals/RequestMessageSnapshotStrategy.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Polly;
+
+namespace Microsoft.Extensions.Http.Resilience.Internal;
+
+/// <summary>
+/// This strategy creates a snapshot of <see cref="HttpRequestMessage"/> before executing the hedging to prevent race conditions when cloning and modifying the message at the same time.
+/// This way, all hedged requests will have an unique instance of the message available from snapshot without the need to access the original one for cloning.
+/// </summary>
+internal sealed class RequestMessageSnapshotStrategy : ResilienceStrategy
+{
+    private readonly IRequestCloner _requestCloner;
+
+    public RequestMessageSnapshotStrategy(IRequestCloner requestCloner)
+    {
+        _requestCloner = requestCloner;
+    }
+
+    protected override async ValueTask<Outcome<TResult>> ExecuteCoreAsync<TResult, TState>(
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state)
+    {
+        var request = context.Properties.GetValue(ResilienceKeys.RequestMessage, null!);
```

Comment:
That ! makes me nervous. Why do we need to break nullability rules here?

Reply:
True, checking for null and throwing exception.